### ISSUE TITLE
Scope Not being applied in middleware

### ISSUE THREAD

**Issue by ghosh** [HipsterJazzbo/Landlord#46]
I added a `ScopeRequestByUser` middleware like so

```php
public function handle($request, Closure $next)
{

    if ($request->ajax() || $request->wantsJson()) {
        Landlord::addTenant('user_id', Auth::guard('api')->id());
    }

    return $next($request);
}
```

Now this works perfectly when I add this to my global middleware stack like this.

```php
protected $middleware = [
    \Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode::class,
    \App\Http\Middleware\ScopeRequestByUser::class,
];
```

However, I need it to be applied only on the api middleware group. When I add to the group in the middleware file it does not work.

```php
    protected $middlewareGroups = [
        'web' => [
          ...
        ],

        'api' => [
            'throttle:60,1',
            'bindings',
            \App\Http\Middleware\ScopeRequestByUser::class,
        ],
    ];
```

I even tried adding it ot the route middleware and then calling it both in the routes file and in the api middleware group. Both of them are not working.

```php
    protected $routeMiddleware = [
        ...
        'scope' => \App\Http\Middleware\ScopeRequestByUser::class,
    ];
```

Not sure, what is going wrong here.

**Comment by dschniepp** [comment]
@ghosh did you find a solution?

**Comment by ghosh** [comment]
@dschniepp Not really, I just ended up using it in the Global Middleware Stack. Hoping this issue can be resolved soon.

**Comment by HipsterJazzbo** [comment]
When you say, "does not work", what do you mean? Tenant are not scoped correctly? At a guess I'd say that the call to `Auth::guard('api')->id()` is returning null for some reason when you register your middleware non-globally?

**Comment by dschniepp** [comment]
When the middleware for setting the tenant will be applied in $middleware everything works as expected and the scoping will be run after that. But if the middleware will be registered in some sort of $routeMiddleware the scope will be applied before the middleware.

**Comment by ghosh** [comment]
I replaced the call to `Landlord::addTenant('user_id', Auth::guard('api')->id());` by `Log::info(Auth::guard('api')->id());` in the `ScopeRequestByUser` middleware and it successfully logs the user id.

This works _regardless_ of weather the middleware is called in the global middleware or routeMiddleware. So the `Auth::guard('api')->id()` is defenetly not returning null.

**Comment by HipsterJazzbo** [comment]
Maybe it's got something to do with one happening before package registration? I'm really stumped here. I'm using Landlord in probably 5 or 6 different middlewares of various types and it works fine.

**Comment by HipsterJazzbo** [comment]
This isn't the same issue as #33 is it?

**Comment by ghosh** [comment]
Hmm...Route Model Binding doesn't seem to be the issue.

Within my routes file, I have this:-
```php
Route::get('/projects', 'ProjectController@all');
```

and the corresponding method in the controller:-
```php
public function all() {
  return Project::all();
}
```

The above does not work. 
However, if I return from within the routes file itself, it seems to work:-
```php
Route::get('/projects', function () {
  return Project::all();
});
```
This works! Not sure whats happening here.

**Comment by torkiljohnsen** [comment]
I am having a similar issue. My problem seems to be the timing of when the tenant is being registered, because some models are being booted before the request is run through the middleware.

From what I can tell, this is the order of things:

1. Kernel runs `sendRequestThroughRouter`
1. In this method, `bootstrap()` is run first
1. This in turn registers all service providers (calls the register method on these). If you are registering any bindings here that includes models (like a repository class), models will be booted now. At this point though, the middleware has not been run yet, so no tenant is registered yet, hence setting up the global scope here fails…
1. After this, the request is run through the middleware where you are setting your tenant. Any models booted after this point will work as expected. Models that were already booted will not have the global scope set.

**Comment by torkiljohnsen** [comment]
I am thinking that the tenant needs to be registered at an earlier point.

Perhaps in the kernel, something like this (this is untested):

```
public function bootstrap()
{
    $app = $this->getApplication();
    $app->beforeBootstrapping(RegisterProviders::class, function () use ($app) {
        // get request with $app->make('request') and set tenant
    });

    parent::bootstrap();
}

**Comment by torkiljohnsen** [comment]
Turns out that loading tenants from the database before the DatabaseServiceProvider has booted can be difficult. 

So a few of other suggestions:

1) Set the tenant in the boot method of a `TenantServiceProvider` and make sure that this provider is loaded before all the other service providers. Hardcoding it like this does not feel quite right. It should be done in some middleware, so that we can control it on a route basis and make sure it happens only on HTTP requests.

2) Make sure all service providers with bindings are deferred, or move bindings to SPs that can be deferred. This can be a lot of fuzz…

3) Perhaps have the `BelongsToTenants` trait override Model::getGlobalScopes method and make sure that the global scope is set every time the global scopes are requested?

**Comment by torkiljohnsen** [comment]
So, the real root of the problem was found finally.

We had set up `observers` in the `boot` method of a service provider, which caused implicated models to boot before middleware was run. There was also another issue that made the models boot prematurely. All resolved now :)

We also had to move the middleware we had set up to register the request tenant from the middlewareGroup to the global middleware in the kernel.

**Comment by dschniepp** [comment]
@torkiljohnsen but registering in a `$middlewareGroup` didn't work or?

**Comment by torkiljohnsen** [comment]
I haven't dug completely down through the call stack to see in what order things are happening in here, but the Kernel seems to be handling running the request through the global `$middleware` and then the router is responsible for running the request through the `$routeMiddleware` and `$middlewareGroups`.

The whole point here is to have the middleware run (and tenant ID set) BEFORE we attempt to set the global scope in models (meaning before any model is booted).

In my case, that meant having to move the middleware to the global `$middleware`. Having it in the middleware handled by the router fails.

**Comment by torkiljohnsen** [comment]
Also: It seems like I will also have to rewrite some code to not use model observers, as that possibly makes models boot when the service providers are booted, before the request is pushed through the global middleware.

**Comment by dschniepp** [comment]
There seems to be a change the implementation between 1.x and 2.x, because 1.x works.

**Comment by torkiljohnsen** [comment]
We changed our approach a bit. The issue is with booting models before the middleware has run. Registering model observers will boot the model, since `Model::observe` will do `new static;`, which calls the constructor and boots the model.

So now I am registering observers like this:

```php
Event::listen('eloquent.booted: ' . User::class, function () {
    User::observe(UserObserver::class);
});
```

This ensures that the observer isn't added until after the User model has been booted by some other process.

Also added an issue for this, can hopefully spark a discussion: https://github.com/laravel/framework/issues/16715

**Comment by themsaid** [comment]
I think the root of the problem is the order in which different components are booted, I believe @torkiljohnsen has the best solution for his issue.

As for this issue, I'm not sure how Landlord works but I assume the issue is related to @torkiljohnsen's, for this please try to see if ELoquent is booted before the request hits the middleware by any mean,  and if the package requires different behaviour.

**Comment by MrRio** [comment]
This is an issue I'm running up against. Where would I add this Event::listen code?

**Comment by torkiljohnsen** [comment]
In a service provider's register method for instance.

**Comment by r3donnx** [comment]
It does work, then I add middleware to the global $middleware. The only problem now, is that in global middleware I can't get authenticated user, it returns null. So I am unable to get users tenant_id.

**Comment by felipeweb11** [comment]
My temporary solution (not the best) was to extend trait, and then I added the tenant after getting the TenantManager from laravel container.

```php

namespace App\Traits;

use HipsterJazzbo\Landlord\BelongsToTenants as BaseBelongsToTenants;
use HipsterJazzbo\Landlord\TenantManager;
use Illuminate\Database\Eloquent\Model;

trait BelongsToTenants
{
    use BaseBelongsToTenants;

    public static function bootBelongsToTenants()
    {
        // Grab our singleton from the container
        static::$landlord = app(TenantManager::class);

        //My temporarily approach
        static::$landlord->addTenant('workspace_id', currentWorkspace()->id);

        // Add a global scope for each tenant this model should be scoped by.
        static::$landlord->applyTenantScopes(new static());

        // Add tenantColumns automatically when creating models
        static::creating(function (Model $model) {
            static::$landlord->newModel($model);
        });
    }

}

**Comment by HipsterJazzbo** [comment]
Here's my understanding:

You can't scope by `Auth::user()` in a global middleware, because the session's not started yet (at least in 5.3+).

Also, until just now, Landlord would accept a NULL user id and carry on happily, which is now fixed in 2.0.6.

Lastly, using Model Observers causes the model's boot sequence to happen before tenants have been added, although any queries done after this point should still work, as the tenants that are set are not actually applied to the model til a query is run.

Does that all seem accurate?

**Comment by Wipsly** [comment]
Any solution?

**Comment by kmcluckie** [comment]
It's not only using model observers that cause the model's boot sequence to happen before tenants have been added - I was using dependency injection to instantiate a model repository, and that also caused the model to boot before the *route* middleware could run and add a tenant. If I added the middleware to the global list (as well as the session and cookie middlewares so I could access the logged in user) then the tenant would be set first. However, I don't want it set on all routes.

The solution was using the suggestion in #61. My middleware uses `clearBootedModels` just before setting the tenant id:

    public function handle($request, Closure $next)
    {
        $user = app('Dingo\Api\Auth\Auth')->user();

        \Illuminate\Database\Eloquent\Model::clearBootedModels();

        foreach($user->tenantIds() as $name => $id)
        {
            Landlord::addTenant($name, $id);
        }

        return $next($request);
    }
